transport/tcp/serve_tcp_ipmap.go

```diff
@@ -60,38 +46,62 @@ func (m *ipMap) Get(ipAddr *net.IPAddr) (string, error) {
 			return e.ident, nil
 		}
 		if e.subnet != nil {
-			contains := e.subnet.Contains(ipAddr.IP)
-			if contains {
-				if e.ident == "willbeclientip" {
-					return ipAddr.IP.String(), nil
-				}
-				return e.ident + ipAddr.IP.String(), nil
+			if e.subnet.Contains(ipAddr.IP) {
+				return strings.Replace(e.ident, "*", ipAddr.IP.String(), 1), nil
 			}
 		}
 	}
 	return "", errors.Errorf("no identity mapping for client IP: %s%%%s", ipAddr.IP, ipAddr.Zone)
 }
 
-func handleIPV6(input string) (net.IP, *net.IPNet, string) {
+func newiPMapEntry(input string, ident string) (*ipMapEntry, error) {
 	ip := input
 	var zone string
+	validationString := ident
 	if strings.Contains(input, "%") {
 		ip = strings.Split(input, "%")[0]
 		zone = strings.Split(input, "%")[1]
 	}
 
-	_, network, err := net.ParseCIDR(ip)
+	_, subnet, err := net.ParseCIDR(ip)
 	if err != nil {
 		parsedIP := net.ParseIP(ip)
 		if parsedIP == nil {
-			return nil, nil, ""
+			return nil, errors.Wrapf(err, "invalid client address for IP %q", ip)
+		}
+
+		if err := transport.ValidateClientIdentity(validationString); err != nil {
+			return nil, errors.Wrapf(err, "invalid client identity for IP %q", ip)
 		}
-		return parsedIP, nil, zone
+		return &ipMapEntry{
+			ip:    parsedIP,
+			zone:  zone,
+			ident: ident,
+		}, nil
 	}
-	return nil, network, zone
+
+	if strings.Count(ident, "*") != 1 {
+		err = fmt.Errorf("CIDRs require 1 IP placeholder")
+		return nil, errors.Wrapf(err, "invalid client identity for IP %q", ip)
+	}
+	if len(ident) == 1 {
+		validationString = strings.Replace(ident, "*", "pass", 1)
+	} else {
+		validationString = strings.Replace(ident, "*", "", 1)
+	}
+
+	if err := transport.ValidateClientIdentity(validationString); err != nil {
```

Comment:
I think it would be best to have the validation string be the maximum-length of an address's `.String()`. perhaps by setting all bits in the address to 1?

About the `zone`: the `%` character is not allowed in a dataset path component (see zfs function `int
zfs_component_namecheck(const char *path, namecheck_err_t *why, char *what)`).

The upcoming big-feature branch #259  brings in better dataset path validation, so we'll start using that in `transport.ValidateClientIdentity`.
=> we will need a way to deal with zones (perhaps detect and forbid them? or only allow them in full mappings?)
==> I'll leave that up to you, just make sure to turn your decision into a test case (for both positive and negatve variants).

Reply:
The current behavior allows for people to use a Zone, and it will be ACL'd against but their is no representation of this in the dataset path.  I imagine we could just do what DNS does here and in the path convert to "-".  It's not a forbidden char. ```const FORBIDDEN = "@#|\t<>*"```  I'll wire something up and see if we like it or not.   This would be handy if someone was doing the same ipv6 subnet ranges but with different zones.  They could prefix "web-eth0-*  but we could just handle that for them.   Probably a better user experience, and is more consistent.

Reply:
"I think it would be best to have the validation string be the maximum-length of an address's .String(). perhaps by setting all bits in the address to 1?"

I'm not following ( I'm kind of a noob. ),  isn't the maximum length of the validationString dependent on the identity provided in the config?  We already know that IP addresses coming in are valid. We must verify that "subnet-one-two-" of  user input "subnet-one-two-*"  validates cleanly.   
```
				"10.1.2.3":    "specific-host",
				"10.1.2.0/24": "subnet-one-two-*",
				"10.1.1.0/24": "subnet-one-one-*",
				"10.1.0.0/24": "subnet-one-zero-*",
```

Speaking of that,  I'll add some tests to assert that all of the forbidden chars are actually forbidden.

Reply:
What I mean is that _for upfron validation when parsing the config input_ we should use the `identity-with-*` from the config, replace `*` with the longest-possible expansion of the left-hand side (`255.255.255.255` for IPv4, `2001:0db8:0000:0000:0000:8a2e:0370:7334` or all `ffff`s for IPv6, plus a maximum zone length if there is any), and feed the result of that expansion to the validation function.

Reply:
> Speaking of that, I'll add some tests to assert that all of the forbidden chars are actually forbidden.

As stated: the actual ZFS implementation (unlike what we do in zrepl's current master branch, see #259 for changes to that) is a whitelist rather than a blacklist => the ZFS function `int zfs_component_namecheck(const char *path, namecheck_err_t *why, char *what)` is authoritative, the implementation in zrepl's current master is incomplete. #259 reimplements `zfs_component_namecheck` in Go, and we'll switch to #259 's impl once that is merged)

Reply:
Ahhh,  that makes more sense now.  Thanks for explaining that.  I didn't take into consideration that a zfs dataset has a max size of 255.    Should we also validate that "root_fs"+"ident"+"ip"+"-zone" <= 255?  I don't think (didn't see) where zfs.NewDatasetPath checks this.

Reply:
I added the following test,  and it passed. (258 chars)
				"10.1.4.0/24": "zwE0tyDzO5mouG3yfB8yCepc6day3ShQzWR2ZSG6J6d2dcBYaWKepCzp9aLtrNg8KoamJM0liqFtZdJJx4L9WUabqEtFy2G45EClbNHhOYxBKCV3shkDi6xtGJVRUOvPhINjLh95uMyHdFvDjBhYkhnobzk64fRgtcy8A4Ad2EorZdKUJ73IrAMNcZFUjuURgHmw9ZBSjlLjIb7vwrjFjI8DWT8xwNu6c3Jcdu7ELihb7t7y4gF254nqDeMESLa1-*",

Reply:
> Should we also validate that "root_fs"+"ident"+"ip"+"-zone" <= 255? I don't think (didn't see) where zfs.NewDatasetPath checks this.

That'll be the job of `transport.ValidateClientIdentity`.
#259 will bring `ComponentNamecheck`, which is what it should use for that purpose.
I assume I'll merge this PR after #259 , so if we rebase this PR after #259 is merged, we can fix that then. I'll add a TODO.

Reply:
(Keep above test in the test suite anyways!)

Reply:
So besides this last validation,  everything else look ok?    How would you look like to move forward,  should we wait until 259 is merged,  and then handle that last little bit?   It look's like TCPListenerFactoryFromConfig would have needed the root_fs passed in also,  unless I'm missing something.    Is there a simpler way to get that piece of config from the job?    Thanks again for review and help.